spring-cloud-gcp-samples/spring-cloud-gcp-pubsub-bus-config-sample/spring-cloud-gcp-pubsub-bus-config-sample-test/src/test/java/com/example/LocalSampleAppIntegrationTest.java

```diff
@@ -0,0 +1,202 @@
+import static org.assertj.core.api.Assertions.fail;
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assume.assumeThat;
+
+/**
+ * Integration test for the config client with local config server.
+ *
+ * <p>Because the test brings in config server/config client source files into a single project,
+ * config server bootstrap need to be suppressed in the config client app instance.
+ *
+ * @author Elena Felder
+ *
+ * @since 1.2
+ */
+public class LocalSampleAppIntegrationTest {
+
+	static final String CONFIG_DIR = "/tmp/config";
+
+	static final String CONFIG_FILE = CONFIG_DIR + "/application.properties";
+
+	static final String INITIAL_MESSAGE = "initial message";
+
+	static final String UPDATED_MESSAGE = "updated message";
+
+	final RestTemplate restTemplate = new RestTemplate();
+
+	BufferedReader configServerOutput;
+
+	BufferedReader configClientOutput;
+
+	Process configServerProcess;
+
+	Process configClientProcess;
+
+	@BeforeClass
+	public static void prepare() throws Exception {
+		assumeThat(
+			"PUB/SUB integration tests are disabled. Use '-Dit.pubsub=true' to enable.",
+			System.getProperty("it.pubsub"), is("true"));
+
+		Files.createDirectories(Paths.get(CONFIG_DIR));
+	}
+
+	@Test
+	public void testSample() throws Exception {
+
+		writeMessageToFile(INITIAL_MESSAGE);
+
+		startConfigServer();
+
+		waitForLogMessage(this.configServerOutput, "Monitoring for local config changes: [" + CONFIG_DIR + "]");
+		waitForLogMessage(this.configServerOutput, "Started PubSubConfigServerApplication");
+		assertConfigServerValue(INITIAL_MESSAGE);
+
+		startConfigClient();
+
+		waitForLogMessage(this.configClientOutput, "Located property source");
+		waitForLogMessage(this.configClientOutput, "Started PubSubConfigApplication");
+		assertConfigClientValue(INITIAL_MESSAGE);
+
+		writeMessageToFile(UPDATED_MESSAGE);
+		waitForLogMessage(this.configServerOutput, "Refresh for: *");
+		assertConfigServerValue(UPDATED_MESSAGE);
+
+		waitForLogMessage(this.configClientOutput, "Keys refreshed [example.message]");
+		assertConfigClientValue(UPDATED_MESSAGE);
+	}
+
+	@AfterClass
+	public static void tearDown() throws Exception {
+
+		Path configFile = Paths.get(CONFIG_FILE);
+		if (Files.exists(configFile)) {
+			Files.delete(configFile);
+		}
+
+		Path configDir = Paths.get(CONFIG_DIR);
+		if (Files.exists(configDir)) {
+			Files.delete(configDir);
+		}
+
+	}
+
+	@After
+	public void closeResources() throws IOException {
+
+		if (this.configServerOutput != null) {
+			this.configServerOutput.close();
+		}
+
+		if (this.configClientOutput != null) {
+			this.configClientOutput.close();
+		}
+
+		if (this.configServerProcess != null) {
+			this.configServerProcess.destroy();
+		}
+
+		if (this.configClientProcess != null) {
+			this.configClientProcess.destroy();
+		}
+
+	}
+
+	private void startConfigServer() throws IOException {
+		ProcessBuilder serverBuilder = new ProcessBuilder("mvn", "spring-boot:run",
```

Comment:
`spring-cloud-gcp-integration-pubsub-sample-test` uses an in-process approach. Anything wrong with that approach?

Reply:
It's fine for the Pub/Sub integration sample, but it's not ideal for the config server sample because config server and config client need different dependencies.
It is possible to make work with an in-process approach, and I did that, but it required using custom several properties to suppress config server functionality in the client, which means that instead of testing the sample, we are testing a completely different scenario (since most of the sample IS configuration).

Here is what the client startup looked like in-process. By the time I was done explaning what deviations from default behavior were needed, it was clear that the sample was not being tested as is.
```
SpringApplicationBuilder
	configClient = new SpringApplicationBuilder(PubSubConfigApplication.class)
	.properties("server.port=8081",
		// suppress config server behavior.
		"spring.cloud.config.server.bootstrap=false",
		// re-enable config client behavior.
		"spring.cloud.config.enabled=true",
		"spring.cloud.config.watch.enabled=true",
		// Suppress Git validation configured through spring.provides in config server module.
		"spring.profiles.active=native"
);

```

Reply:
I see. Thanks for the clarification.